pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,220 @@
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodeRef{Client: mgr.GetClient(), scheme: mgr.GetScheme(), recorder: mgr.GetRecorder(controllerName)}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("noderef-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Machines.
+	return c.Watch(&source.Kind{Type: &v1alpha1.Machine{}}, &handler.EnqueueRequestForObject{})
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeRef{}
+
+// ReconcileNodeRef reconciles a Machine object.
+type ReconcileNodeRef struct {
+	client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// Reconcile watches Machines.
+// +kubebuilder:rbac:groups=,resources=secrets,verbs=get;list;watch
+func (r *ReconcileNodeRef) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	klog.Infof("Reconcile request for Machine %q in namespace %q", request.Name, request.Namespace)
+	ctx := context.Background()
+
+	// Fetch the NodeRef instance
+	machine := &v1alpha1.Machine{}
+	err := r.Get(ctx, request.NamespacedName, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).Infof("Machine %q in namespace %q is not found, won't reconcile", machine.Name, machine.Namespace)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		klog.V(2).Infof("Machine %q in namespace %q has been deleted, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine doesn't already have a NodeRef.
+	if machine.Status.NodeRef != nil {
+		klog.V(2).Infof("Machine %q in namespace %q already has a NodeRef, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a cluster label.
+	if machine.Labels[v1alpha1.MachineClusterLabelName] == "" {
+		klog.V(2).Infof("Machine %q in namespace %q doesn't specify %q label, won't reconcile", machine.Name, machine.Namespace,
+			v1alpha1.MachineClusterLabelName)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, retrying later", machine.Name, machine.Namespace)
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	result, err := r.reconcile(ctx, machine)
+	if err != nil {
+		klog.Errorf("Failed to assign NodeRef to Machine %q: %v", request.NamespacedName, err)
+		r.recorder.Eventf(machine, apicorev1.EventTypeWarning, "NodeRefReconcileError", "%v", err)
```

Comment:
You can use `Event` instead of `Eventf` if you don't have any string formatting, and `err.Error()` instead of the `%v`?

Reply:
And maybe something like `FailedSetNotRef` would be more in line with some of the other events, such as `SuccessfulCreate`?